lib/SILOptimizer/Differentiation/PullbackCloner.cpp

```diff
@@ -400,12 +422,59 @@ class PullbackCloner::Implementation final
     }
     /// If adjoint value is concrete, it is already materialized. Store it in
     /// the destination address.
-    case AdjointValueKind::Concrete:
+    case AdjointValueKind::Concrete: {
       auto concreteVal = val.getConcreteValue();
       builder.emitStoreValueOperation(loc, concreteVal, destAddress,
                                       StoreOwnershipQualifier::Init);
       break;
     }
+    case AdjointValueKind::AddElement: {
+      auto baseAdjoint = val;
+
+      // Current adjoint may be made up of layers of `AddElement` adjoints.
+      // We can iteratively gather the list of elements to add instead of making
+      // recursive calls to `materializeAdjointIndirect`.
+      SmallVector<AddElementValue *, 4> addEltAdjValues;
+
+      do {
+        auto addElementValue = baseAdjoint.getAddElementValue();
```

Comment:
Maybe add some assertions here? After all, all these adoints must be either tuple-type or struct-like.